runtime/Cpp/runtime/src/support/InlinedVector.h

```diff
@@ -0,0 +1,346 @@
+/* Copyright (c) 2012-2021 The ANTLR Project. All rights reserved.
+ * Use of this file is governed by the BSD 3-clause license that
+ * can be found in the LICENSE.txt file in the project root.
+ */
+
+#pragma once
+
+#include <algorithm>
+#include <cassert>
+#include <cstddef>
+#include <iterator>
+#include <limits>
+#include <memory>
+#include <stdexcept>
+#include <type_traits>
+#include <utility>
+
+namespace antlrcpp {
+
+  template <typename T>
+  struct InlinedVectorRemoteStorage final {
+    T *data;
+    size_t size;
+  };
+
+  template <typename T,
+            size_t N = (sizeof(InlinedVectorRemoteStorage<T>) / sizeof(T)),
+            typename Allocator = std::allocator<T>>
+  class InlinedVector;
+
+  template <typename T, size_t N, typename Allocator>
+  class InlinedVector final {
+  private:
+    union Storage final {
+      Storage() {}
+
+      InlinedVectorRemoteStorage<T> remote;
+      T local[N];
+    };
+
+  public:
+    using value_type = T;
+    using allocator_type = Allocator;
+    using size_type = size_t;
+    using difference_type = ptrdiff_t;
+    using reference = value_type&;
+    using const_reference = const value_type&;
+    using pointer = value_type*;
+    using const_pointer = const value_type*;
+    using iterator = pointer;
+    using const_iterator = const_pointer;
+    using reverse_iterator = std::reverse_iterator<iterator>;
+    using const_reverse_iterator = std::reverse_iterator<const_iterator>;
+
+    InlinedVector() = default;
+
+    InlinedVector(const InlinedVector<T, N, Allocator> &other) {
+      copy_from(other);
+    }
+
+    InlinedVector(InlinedVector<T, N, Allocator> &&other) {
+      move_from(std::forward<InlinedVector<T, N, Allocator>>(other));
+    }
+
+    ~InlinedVector() {
+      destroy();
+    }
+
+    InlinedVector<T, N, Allocator>& operator=(const InlinedVector<T, N, Allocator> &other) {
+      if (this != std::addressof(other)) {
```

Comment:
To reduce the cyclomatic complexity and improve readability, you could return early after the aliasing check:

```c++
if (this == std::addressof(other)) return *this;
if (capacity()....
```